### PR TITLE
API-1835: separate mutations into proper targets based on output-resources

### PR DESCRIPTION
**Pull request by deads2k**
This uses `output-resources` to determine which type of cluster the various targets belong to.  It also adds checking to be sure that every output from `apply-configuration` has a matching `output-resources`.

/assign @p0lyn0mial @bertinatto

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -29,6 +29,7 @@ $(call build-image,multi-operator-manager,$(IMAGE_REGISTRY)/ocp/4.18:multi-opera
 
 clean:
 	$(RM) multi-operator-manager
+	$(RM) -rf test-output
```
Comment (deads2k) [inline Makefile:32]: @bertinatto  turns out this is important for integration testing

File: pkg/library/libraryapplyconfiguration/client_mutations.go [hunk 2]
```diff
@@ -22,7 +24,9 @@ func (s *clientBasedClusterApplyResult) Requests() MutationActionReader {
 	return s.mutationTracker
 }
 
-func NewApplyConfigurationFromClient(mutationTracker *manifestclient.AllActionsTracker[manifestclient.TrackedSerializedRequest]) *applyConfiguration {
+func NewApplyConfigurationFromClient(
```
Comment (p0lyn0mial) [inline pkg/library/libraryapplyconfiguration/client_mutations.go:27]: This function appears to be assigning the mutationTracker, which holds all serialized requests, to all types of clusters. 

However, I don't see where the mapping of serialized requests to specific cluster types is being done.

Comment (deads2k) [reply pkg/library/libraryapplyconfiguration/client_mutations.go:27]: > However, I don't see where the mapping of serialized requests to specific cluster types is being done.

15 lines, down in `FilterAllDesiredMutationsGetter` https://github.com/openshift/multi-operator-manager/pull/16/files#diff-7c6bf328fe10881be715bb2460cc7d1628316f4aa43b99438274b53d270768e0R43 called from the command: https://github.com/openshift/multi-operator-manager/pull/16/files#diff-817f4737e0c52096f9428fd232ea539f40b2885a60d93b387bfd86bc1c654a66R56

This way each operator doesn't have to get it just right.

Comment (deads2k) [reply pkg/library/libraryapplyconfiguration/client_mutations.go:27]: and the outputs themselves are here: https://github.com/openshift/cluster-authentication-operator/blob/master/pkg/cmd/mom/output_resources_command.go#L16-L43

File: pkg/library/libraryapplyconfiguration/apply_configuration.go [hunk 3]
```diff
@@ -35,21 +37,41 @@ var (
...
-		desiredMutationsGetter := s.MutationsForClusterType(clusterType)
+		desiredMutationsGetter := allDesiredMutationsGetter.MutationsForClusterType(clusterType)
 		switch {
 		case desiredMutationsGetter == nil:
 			errs = append(errs, fmt.Errorf("mutations for %q are required even if empty", clusterType))
 		case desiredMutationsGetter.GetClusterType() != clusterType:
 			errs = append(errs, fmt.Errorf("mutations for %q reported type=%q", clusterType, desiredMutationsGetter.GetClusterType()))
 		}
+
+		allMutationRequests = append(allMutationRequests, desiredMutationsGetter.Requests().AllRequests()...)
```
Comment (p0lyn0mial) [inline pkg/library/libraryapplyconfiguration/apply_configuration.go:57]: I think that `allMutationRequests` contains already filtered output. 

`AllRequests` does: https://github.com/openshift/multi-operator-manager/blob/3db370e102a9182e54ae8cf88002f575e2823b96/pkg/library/libraryapplyconfiguration/client_mutations.go#L110

File: pkg/library/libraryapplyconfiguration/client_mutations.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/library/libraryapplyconfiguration/client_mutations.go:27]: OK, thanks for the pointers.

File: pkg/library/libraryapplyconfiguration/apply_configuration.go [hunk 4]
```diff
@@ -35,21 +37,41 @@ var (
...
+	combinedList := &libraryoutputresources.ResourceList{}
+	combinedList.ExactResources = append(combinedList.ExactResources, allAllowedOutputResources.ConfigurationResources.ExactResources...)
+	combinedList.ExactResources = append(combinedList.ExactResources, allAllowedOutputResources.ManagementResources.ExactResources...)
+	combinedList.ExactResources = append(combinedList.ExactResources, allAllowedOutputResources.UserWorkloadResources.ExactResources...)
+	combinedList.GeneratedNameResources = append(combinedList.GeneratedNameResources, allAllowedOutputResources.ConfigurationResources.GeneratedNameResources...)
+	combinedList.GeneratedNameResources = append(combinedList.GeneratedNameResources, allAllowedOutputResources.ManagementResources.GeneratedNameResources...)
+	combinedList.GeneratedNameResources = append(combinedList.GeneratedNameResources, allAllowedOutputResources.UserWorkloadResources.GeneratedNameResources...)
+	filteredMutationRequests := FilterSerializedRequests(allMutationRequests, combinedList)
+
+	if unspecifiedOutputResources := manifestclient.DifferenceOfSerializedRequests(allMutationRequests, filteredMutationRequests); len(unspecifiedOutputResources) > 0 {
```
Comment (p0lyn0mial) [inline pkg/library/libraryapplyconfiguration/apply_configuration.go:69]: So the second call to `ValidateAllDesiredMutationsGetter` will contain filtered mutations. So this function will not find a diff. Is this expected?

https://github.com/openshift/multi-operator-manager/blob/3db370e102a9182e54ae8cf88002f575e2823b96/pkg/library/libraryapplyconfiguration/options.go#L57

File: hack/test-operator-integration.sh [hunk 5]
```diff
@@ -4,4 +4,5 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+rm -rf ./test-output
```
Comment (p0lyn0mial) [inline hack/test-operator-integration.sh:7]: why not to add the `clean` target to https://github.com/openshift/multi-operator-manager/blob/3db370e102a9182e54ae8cf88002f575e2823b96/Makefile#L46 target ?

File: pkg/library/libraryapplyconfiguration/options.go [hunk 6]
```diff
@@ -5,42 +5,60 @@ import (
...
+
+	// also validate the raw results because filtering may have eliminated "bad" output.
+	if err := ValidateAllDesiredMutationsGetter(result, allAllowedOutputResources); err != nil {
+		errs = append(errs, err)
+	}
+
+	// now filter the results and check them
+	filteredResult := FilterAllDesiredMutationsGetter(result, allAllowedOutputResources)
+	if err := ValidateAllDesiredMutationsGetter(filteredResult, allAllowedOutputResources); err != nil {
 		errs = append(errs, err)
```
Comment (p0lyn0mial) [inline pkg/library/libraryapplyconfiguration/options.go:58]: could we add some context to this error so that we know it failed on already filtered data?

**Comment by deads2k** [comment]
the reset promised as IOUs

**Pull request merged**